_posts/2019-02-25-practice-oop-to-front-end-universal-state-module.md

```diff
@@ -0,0 +1,291 @@
+
+Let's start with a typical Redux example of a counter:
+
+```js
+import { createStore } from 'redux';
+
+function counter(state = 0, action) {
+  switch (action.type) {
+    case 'INCREMENT':
+      return state + 1;
+    case 'DECREMENT':
+      return state - 1;
+    default:
+      return state;
+  }
+}
+
+const store = createStore(counter)
+
+store.dispatch({ type: 'INCREMENT' })
+store.dispatch({ type: 'DECREMENT' })
+```
+
+USM supports Redux, MobX, Vuex and Angular. It provides `usm`, `usm-redux`, `usm-mobx` and `usm-vuex` packages. Here is the same counter using `usm-redux`:
+
+```js
+import Module, { state, action } from 'usm-redux';
+
+class Counter extends Module {
+  @state count = 0;
+
+  @action
+  increase(state) {
+    state.count += 1;
+  }
+
+  @action
+  decrease(state) {
+    state.count -= 1;
+  }
+}
+
+const counter = Counter.create();
+
+counter.increase();
+counter.decrease();
+```
+
+We want the Redux-based model to be more intuitive and concise, just like the OO example of ES6+ 'class' counter mentioned above, if such an OO paradigm is also a generic state model, a better unified state library encapsulation, This will undoubtedly lead to a more flexible and friendly development experience for developers (of course, include easy to read/maintain, etc.).
+
+USM has four sub-packages, which are `usm`, `usm-redux`, `usm-mobx` and `usm-vuex`. In this example, it mainly uses `usm-redux`. `usm-redux` is mainly based on Redux and [Immer](https://github.com/mweststrate/immer). Immer is mainly used in mutable modifying created to get immutable state.
+
+The following is the connector section using `usm-redux`:
+
+```js
+// index.js
+export const counter = Counter.create();
+
+ReactDOM.render(
+  <Provider store={counter.store}>
+    <App />
+  </Provider>,
+  document.getElementById('root')
+);
+```
+
+```js
+// app.js
+import { connect } from 'react-redux';
+import { counter } from './';
+
+export default connect(
+  state => ({ ount: state.count })
+)( props => 
+  <div>
+    <button onClick={() => counter.increase()}>+</button>
+    {props.count}
+    <button onClick={() => counter.decrease()}>-</button>
+  </div>
+);
+```
+
+And here is the same counter `View` connector using `usm-mobx`:
+
+```js
+// index.js
+
+export const counter = Counter.create();
+
+ReactDOM.render(
+  <App />,
+  document.getElementById('root')
+);
+```
+
+
+```js
+// app.js
+import { observer } from 'mobx-react';
+import { counter } from './';
+
+export default observer(() =>
+  <div>
+    <button onClick={() => counter.increase()}>+</button>
+    {counter.count}
+    <button onClick={() => counter.decrease()}>-</button>
+  </div>
+);
+```
+
+In the `usm-redux` and `usm-mobx` examples, in addition to some differences in the View connector, the state module part is in accord, and this is the universal state module that we had proposed.
```

Comment:
The use of `usm-redux` and `usm-mobx` to connect with `react-redux` and `mobx-react` respectfully demonstrated that the core implementations of the state module is the same even when the connectors used is different. This is the core principle of the Universal State Module that we propose.